nion/eels_analysis/BackgroundModel.py

```diff
@@ -119,8 +119,41 @@ def __unused_perform_fit(self, xs: numpy.ndarray, ys: numpy.ndarray, fs: numpy.n
         series = typing.cast(typing.Any, numpy.polynomial.polynomial.Polynomial.fit(xs, transform_data(ys), self.deg))
         return untransform_data(series(fs))
 
+class TwoAreaBackgroundModel(AbstractBackgroundModel):
+
+    def __init__(self, background_model_id: str, params_func = None, model_func = None, title: str = None):
+        super().__init__(background_model_id, title)
+        self.model_func = model_func
+        self.params_func = params_func
+
+
+    def _perform_fits(self, xs: numpy.ndarray, yss: numpy.ndarray, fs: numpy.ndarray) -> numpy.ndarray:
+        if not len(xs) % 2 == 0:
+            yss = yss[...,:-1]
+            xs = xs[:-1]
+        half_interval = len(xs) // 2
+        areas_1 = numpy.trapz(yss[...,:half_interval], xs[:half_interval], axis=1)
+        areas_2 = numpy.trapz(yss[...,half_interval:], xs[half_interval:], axis=1)
+        x_start = xs[0]
+        x_center = xs[half_interval]
+        x_end = xs[-1]
+        p1s, p2s = self.params_func(areas_1,areas_2, x_start, x_center, x_end)
+        series = numpy.empty((len(yss),len(fs)))
+        for idx, (p1, p2) in enumerate(zip(p1s,p2s)):
+            series[idx] = self.model_func(fs,p1,p2)
+        return series
+
+def power_law_params(areas_1: numpy.ndarray, areas_2: numpy.ndarray, x_start: float, x_center: float, x_end: float) -> typing.Tuple[numpy.ndarray, numpy.ndarray]:
+    r = 2 * (numpy.log(areas_1) - numpy.log(areas_2)) / (numpy.log(x_end) - numpy.log(x_start))
+    k = 1 - r
+    A = k * areas_2 / (x_end ** k - x_center ** k)
+    return A, r
+
+def power_law_func(x: numpy.ndarray, A: float, r: float) -> numpy.ndarray:
+    return A*x**-r
```

Comment:
Can this be rewritten to `return A[...,:] * x[...,:] ** -r[...,:]` to avoid the explicit loop on lines 142-143? One thing that might be helpful is to add comments about what dimensions everything will be on input and output. That would help answer this question. Hmm. Maybe `return A[:] * x[...,:] ** -r[:]`?

Reply:
This function as it is can accommodate 1D arrays for all inputs (x, A, and r), but that doesn't give you what you want, because if you have an array of different A values and r values, that gets broadcast element by element with the x array so that each x value is combined with a different A and r value. 

Lets say m is the number of spectra in the SI, and n is the length of the background display interval. 

In the single spectrum case, x is a 1D array of energy values (length n), and A and r are floats, and we output a 1D array of y values (length n).

In the SI case, A and r are 1D arrays of length m. We want to combine each pair of scalar values in A and r with the same 1D array of x values. The output should be a m x n array of y values.

Reply:
I figured out a better way. I almost had this working in a previous version, but I needed to transpose the input array. The function itself is fine as it is. This version takes about 5s with the 512 x 512 SI.

`       p1s, p2s = self.params_func(areas_1, areas_2, x_start, x_center, x_end)
        grid = numpy.transpose(numpy.tile(fs,(len(yss),1)))
        series = typing.cast(typing.Any, numpy.transpose(self.model_func(grid,p1s,p2s)))`